### PR TITLE
(WIP) Fuel accuracy fix for #37943

### LINKED ISSUE #37943: Fuel remaining wildly inaccurate
When 'e'xamining a vehicle you're driving the fuel remaining counter seems to calculate time until empty by using the old time format where one turn = three seconds. There may be other issues with it, but from some limited testing it seems that this accounts for most inaccuracy. 

# Steps To Reproduce
1. Get an electric vehicle and attach a small (e.g. car) battery.
2. Get up to a cruising speed of whatever you like.
3. 'e'xamine and note the time remaining on the clock. Compare the time remaining every few turns. Minutes will tick by over a matter of seconds.

# Versions and configuration

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version
- Game Version: 0.D-12105-gda56791 [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Aftershock [aftershock],
    C.R.I.T Expansion Mod [crt_expansion],
    Magiclysm [magiclysm],
    Makeshift Items Mod [makeshift],
    More Survival Tools [More_Survival_Tools],
    Mythological Replicas [nw_pack],
    Hydroponics [hydroponics],
    Mutant NPCs [mutant_npcs],
    No Rail Stations [No_Rail_Stations],
    Folding Parts pack [deoxymod],
    Vehicle Additions Pack [blazemod],
    Tanks and Other Vehicles [Tanks],
    Roadheader and other mining vehicles [Heavy miners],
    No Fungal Monsters [No_Fungi],
    Beta National Guard Camp [national_guard_camp]
]

### PR DESCRIPTION
**Pull request by jkraybill**
#### Summary
```SUMMARY: Interface "Improve accuracy of remaining fuel time"```

#### Purpose of change
Fixes #37943 

I'm marking this WIP because it still isn't accurate, but it's way more accurate than before and I'm not sure it will ever be.

The idea of this change is to use the actual last-tick fuel consumption data to estimate remaining fuel time. However, I'm not an expert in physics and the formulas were really mind-bending. I think I have the general idea right but I'm not certain and am PR'ing this to get some expert feedback.

#### Describe the solution
Instead of the old calculation, which seemed to be quite far off the mark, I'm using the last-tick fuel consumption number to calculate remaining fuel.

#### Describe alternatives you've considered
I nearly gave up after several hours of reading the code and doing road tests.

#### Testing
I spawned a very long straight road using overmap specials, and then spawned diesel, gas, and electric vehicles. I siphoned the tanks down to < 0.5L for the gas/diesel, and installed small car batteries in the electrics, and then ran them at various speeds and looked at the remaining time.

At the moment, it looks like it is still slightly under-estimating the time remaining for electric vehicles, and it gets numbers that bounce up and down a little for gas/diesel, but it appears to be way closer than before at least.

#### Additional context
I took this on for my first real C++ challenge on this project. I may have bitten off more than I can chew.

### CODE REVIEW

File: src/vehicle.cpp [hunk 1]
```diff
@@ -4451,13 +4429,15 @@ void vehicle::consume_fuel( int load, const int t_seconds, bool skip_electric )
...
         double remainder = fuel_remainder[ ft ];
         amnt_precise_j -= remainder;
 
         if( amnt_precise_j > 0.0f ) {
-            fuel_remainder[ ft ] = drain_energy( ft, amnt_precise_j ) - amnt_precise_j;
+            double actual_drain = drain_energy( ft, amnt_precise_j );
+            inserted.first->second += actual_drain;
+            fuel_remainder[ ft ] = actual_drain - amnt_precise_j;
         } else {
+	        inserted.first->second += amnt_precise_j;
```
Comment (davidpwbrown) [inline src/vehicle.cpp]: indent here to match double tab.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -3492,30 +3492,8 @@ int vehicle::consumption_per_hour( const itype_id &ftype, int fuel_rate_w ) cons
...
-                // otherwise have been division-by-zero
-                debugmsg( "Vehicle unexpectedly has zero acceleration" );
-            } else {
-                amount_pct += 600 * vslowdown / accel;
-            }
-        }
-    }
-    int energy_j_per_mL = fuel.fuel_energy() * 1000;
-    return -amount_pct * fuel_rate_w / energy_j_per_mL;
+    int energy_j_per_L = fuel.fuel_energy() * 1000;
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: fuel.fuel_energy() returns MJ/L
basic dimensional analysis:
1 MJ/L * 1,000,000 J/MJ * 1/1000 L/mL = 1000 J/mL

1 MJ/L * 1,000,000 J/MJ * 1 L/L = 1,000,000 J/L

So mathematically, the old code was correct and the new code is wrong.

File: src/vehicle.cpp [hunk 3]
```diff
@@ -3492,30 +3492,8 @@ int vehicle::consumption_per_hour( const itype_id &ftype, int fuel_rate_w ) cons
...
+    return -3600 * ( fuel_rate_w / energy_j_per_L );
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: What are you even trying to do here?  Okay, maybe reporting nominal fuel consumption at a fixed speed isn't a great solution, but vehicles also don't normally run at 100% of their engine capacity.

File: src/vehicle.cpp [hunk 4]
```diff
@@ -3492,30 +3492,8 @@ int vehicle::consumption_per_hour( const itype_id &ftype, int fuel_rate_w ) cons
...
-        // HACK: engines consume 1 second worth of fuel per turn, even though a turn is 6 seconds
-        if( vslowdown > 0 ) {
-            int accel = acceleration( true, target_v );
-            if( accel == 0 ) {
-                // FIXME: Long-term plan is to change the fuel consumption
-                // computation entirely; for now just warn if this would
-                // otherwise have been division-by-zero
-                debugmsg( "Vehicle unexpectedly has zero acceleration" );
-            } else {
-                amount_pct += 600 * vslowdown / accel;
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: this should probably be `amount_pct += 3600 * vslowdown / accel;` and the comment starting with `// HACK: ` should be removed now that it is no longer true.

File: src/vehicle.cpp [hunk 5]
```diff
@@ -4451,13 +4429,15 @@ void vehicle::consume_fuel( int load, const int t_seconds, bool skip_electric )
         double amnt_precise_j = static_cast<double>( fuel_pr.second ) * t_seconds;
         amnt_precise_j *= load / 1000.0 * ( 1.0 + st * st * 100.0 );
         auto inserted = fuel_used_last_turn.insert( { ft, 0.0f } );
-        inserted.first->second += amnt_precise_j;
         double remainder = fuel_remainder[ ft ];
         amnt_precise_j -= remainder;
 
         if( amnt_precise_j > 0.0f ) {
-            fuel_remainder[ ft ] = drain_energy( ft, amnt_precise_j ) - amnt_precise_j;
+            double actual_drain = drain_energy( ft, amnt_precise_j );
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: Why are you messing with the functions that handle fuel consumption in a PR that is supposed to be about displaying fuel consumption?

If this function is buggy, open a separate issue and a PR to fix to it, don't do it in this one please.

File: src/vehicle.cpp [hunk 6]
```diff
@@ -3492,30 +3492,8 @@ int vehicle::consumption_per_hour( const itype_id &ftype, int fuel_rate_w ) cons
     if( fuel_rate_w == 0 || fuel.has_flag( "PERPETUAL" ) || !engine_on ) {
         return 0;
     }
-    // consume this fuel type's share of alternator load for 3600 seconds
-    int amount_pct = 3600 * alternator_load / 1000;
-
-    // calculate fuel consumption for the lower of safe speed or 70 mph
-    // or 0 if the vehicle is idling
-    if( is_moving() ) {
-        int target_v = std::min( safe_velocity(), 70 * 100 );
```
Comment (mlangsdorf) [inline src/vehicle.cpp]: Thinking about this, almost no one drives at 70 mph, which is a brisk 17 tiles/turn.  Calculating nominal fuel consumption at 40 mph would probably be better.

**Comment by jkraybill** [comment]
Thanks for the feedback @mlangsdorf! I'm going to close this PR and re-open a simpler one that addresses the off-by-6x issue and changes the calculations to 40mph.

**Comment by jkraybill** [comment]
I tried going back and simply fixing the 6x issue and switching to 40mph, and the remaining time estimates were still off by an order of magnitude at least. I'm going to back off of this one and leave it to the experts for now!

**Pull request closed without merging**